proposals/0088-enable-core-bpf-programs.md

```diff
@@ -0,0 +1,143 @@
+`BPFLoader2111111111111111111111111111111111`.
+
+Core BPF programs must only be modified through feature gates.
+
+**Upgrading a core BPF program** shall consist of deploying the modified program
+source code to a new arbitrary address as a **non-upgradeable** BPF program and
+creating a feature gate to, using the runtime, move the modified program in
+place of the existing program. 
+
+Note, since this deployed program will be deployed using the non-upgradeable BPF
+loader (`BPFLoader2111111111111111111111111111111111`), it will consist of only
+a program account, with no additional program _data_ account.
+
+This process must adhere to the following steps:
+
+1. Verifiably build the modified program.
+2. Generate a new keypair for the **source** program.
+3. Deploy the program to the **source** address.
+4. Generate a new keypair for the **feature gate**.
+5. Create a new feature gate for replacing the **target** program with the
+   **source** (modified) program.
+6. Follow the existing process for activating features.
+
+An additional optional field could be added to feature gate issues for the
+**target program** being upgraded.
+
+**Migrating a native program** to a core BPF program shall consist of following
+the exact steps outlined above with the modified BPF version of the program.
+This will effectively replace the program at the target address with its core
+BPF implementation.
+
+Some additional checks should be run when doing a migration instead of an
+upgrade, such as:
+
+- Validating the native program's owner is
+  `NativeLoader1111111111111111111111111111111`
+- Ensuring the core BPF program has the correct state
+
+## Impact
+
+This proposed change would result in a new process for upgrading core programs.
+Although this process will still require feature gates as it does now, the
+change is handled completely differently.
+
+The act of deploying the modified program and using a runtime feature gate to
+move it into the proper account will be an entirely new way of upgrading
+programs that were previously native.
+
+Core contributors will have to become comfortable with this process.
+
+## Security Considerations
+
+This proposal provides a secure means for upgrading core BPF programs - a
+process that will be increasingly valuable as new core BPF programs are created.
+
+However, this change also introduces a critical security consideration
+surrounding the replacement of a core program with the contents of another
+account. This is an extremely sensitive process that must be handled with
+maximum caution.
+
+If a core program is modified incorrectly, or somehow erased during migration,
+it could have immediate and fatal consequences for the network.
+
+## Backwards Compatibility
+
+This feature itself does not directly introduce any breaking changes.
+
+When this proposed process is used to migrate a native program to core BPF, that
+program will be completely backwards compatible.
```

Comment:
It might be nice to point out that the core BPF implementation must be more than backwards compatible, and provide exactly the same results as the current native program.